File: kit/swift/Sources/NerdKeyKit/LicenseState.swift
import Foundation

/// Persisted license state written to license.json.
public struct LicenseState: Codable {
    public var schemaVersion: Int = 1
    public var licenseKey: String
    public var machineId: String
    public var fingerprint: String
    public var lastOnlineCheckAt: Date?
    public var lastOnlineCheckResult: String?
    public var cachedKeygenLicenseId: String?

    public init(
        licenseKey: String,
        machineId: String,
        fingerprint: String,
        lastOnlineCheckAt: Date? = nil,
        lastOnlineCheckResult: String? = nil,
        cachedKeygenLicenseId: String? = nil
    ) {
        self.licenseKey = licenseKey
        self.machineId = machineId
        self.fingerprint = fingerprint
        self.lastOnlineCheckAt = lastOnlineCheckAt
        self.lastOnlineCheckResult = lastOnlineCheckResult
        self.cachedKeygenLicenseId = cachedKeygenLicenseId
    }
}

/// Manages reading and writing license.json from the platform-appropriate location.
public struct LicenseStateStore {
    private let fileURL: URL

    public init(appSlug: String, stateDir: String? = nil) {
        if let override = stateDir {
            let dir = URL(fileURLWithPath: override, isDirectory: true)
            try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
            fileURL = dir.appendingPathComponent("license.json")
        } else {
            let baseDir = LicenseStateStore.defaultBaseDirectory(appSlug: appSlug)
            try? FileManager.default.createDirectory(at: baseDir, withIntermediateDirectories: true)
            fileURL = baseDir.appendingPathComponent("license.json")
        }
    }

    private static func defaultBaseDirectory(appSlug: String) -> URL {
        #if os(macOS)
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        return base.appendingPathComponent("nerdsmiths/\(appSlug)", isDirectory: true)
        #elseif os(Linux)
        let home = ProcessInfo.processInfo.environment["HOME"] ?? "/tmp"
        return URL(fileURLWithPath: "\(home)/.local/share/nerdsmiths/\(appSlug)", isDirectory: true)
        #else
        // Windows fallback (not reachable on Apple/Linux platforms this code targets)
        let appData = ProcessInfo.processInfo.environment["APPDATA"] ?? "/tmp"
        return URL(fileURLWithPath: "\(appData)/nerdsmiths/\(appSlug)", isDirectory: true)
        #endif
    }

    public func load() throws -> LicenseState? {
        guard FileManager.default.fileExists(atPath: fileURL.path) else { return nil }
        let data = try Data(contentsOf: fileURL)
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try decoder.decode(LicenseState.self, from: data)
    }

    public func save(_ state: LicenseState) throws {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        let data = try encoder.encode(state)
        try data.write(to: fileURL, options: .atomic)
    }

    public func delete() {
        try? FileManager.default.removeItem(at: fileURL)
    }

    public var filePath: String { fileURL.path }
}

File: kit/dotnet/NerdKey.Kit/LicenseState.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Runtime.InteropServices;

namespace NerdKey.Kit;

/// <summary>
/// Persisted license state written to license.json.
/// Locations:
///   macOS:   ~/Library/Application Support/nerdsmiths/{appSlug}/license.json
///   Windows: %APPDATA%\nerdsmiths\{appSlug}\license.json
///   Linux:   ~/.local/share/nerdsmiths/{appSlug}/license.json
/// </summary>
public sealed class LicenseState
{
    [JsonPropertyName("schemaVersion")]
    public int SchemaVersion { get; set; } = 1;

    [JsonPropertyName("licenseKey")]
    public string LicenseKey { get; set; } = "";

    [JsonPropertyName("machineId")]
    public string MachineId { get; set; } = "";

    [JsonPropertyName("fingerprint")]
    public string Fingerprint { get; set; } = "";

    [JsonPropertyName("lastOnlineCheckAt")]
    public DateTimeOffset? LastOnlineCheckAt { get; set; }

    [JsonPropertyName("lastOnlineCheckResult")]
    public string? LastOnlineCheckResult { get; set; }

    [JsonPropertyName("cachedKeygenLicenseId")]
    public string? CachedKeygenLicenseId { get; set; }
}

/// <summary>Manages reading and writing license.json.</summary>
public sealed class LicenseStateStore
{
    private readonly string _filePath;

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    public LicenseStateStore(string appSlug, string? stateDirOverride = null)
    {
        string dir;
        if (stateDirOverride is not null)
        {
            dir = stateDirOverride;
        }
        else
        {
            dir = Path.Combine(DefaultBaseDirectory(), "nerdsmiths", appSlug);
        }
        Directory.CreateDirectory(dir);
        _filePath = Path.Combine(dir, "license.json");
    }

    private static string DefaultBaseDirectory()
    {
        if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
        {
            var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
            return Path.Combine(home, "Library", "Application Support");
        }
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
        {
            return Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
        }
        // Linux
        var xdgData = Environment.GetEnvironmentVariable("XDG_DATA_HOME");
        if (!string.IsNullOrEmpty(xdgData)) return xdgData;
        var homeDir = Environment.GetEnvironmentVariable("HOME") ?? "/tmp";
        return Path.Combine(homeDir, ".local", "share");
    }

    public string FilePath => _filePath;

    public LicenseState? Load()
    {
        if (!File.Exists(_filePath)) return null;
        var json = File.ReadAllText(_filePath);
        return JsonSerializer.Deserialize<LicenseState>(json, JsonOptions);
    }

    public void Save(LicenseState state)
    {
        var json = JsonSerializer.Serialize(state, JsonOptions);
        File.WriteAllText(_filePath, json);
    }

    public void Delete()
    {
        if (File.Exists(_filePath))
            File.Delete(_filePath);
    }
}
